File: fdm-yolov8n-vgg16/src/config/ModelConfig.ts
/**
 * 모델 설정 및 클래스 정보 관리
 */

export interface ModelClassInfo {
  id: number;
  name: string;
  nameEn: string;
  description?: string;
  color: string;
  icon: string;
}

export interface ModelConfig {
  yoloClasses: ModelClassInfo[];
  vggClasses: ModelClassInfo[];
  inputTensorSize: number[];
  confidenceThreshold: number;
  iouThreshold: number;
}

/**
 * YOLOv8 증상 클래스
 */
export const YOLO_CLASSES: ModelClassInfo[] = [
  { id: 0, name: 'Bleeding', nameEn: 'Bleeding', color: '#F44336', icon: 'drop' },
  { id: 1, name: 'Corrosion', nameEn: 'Corrosion', color: '#FF9800', icon: '腐蚀' },
  { id: 2, name: 'Tumor', nameEn: 'Tumor', color: '#9C27B0', icon: 'circle-outline' },
  { id: 3, name: 'Ulcer', nameEn: 'Ulcer', color: '#E91E63', icon: 'alert-circle' },
  { id: 4, name: 'EyesSymptom', nameEn: 'Eye Symptom', color: '#2196F3', icon: 'eye' },
];

/**
 * VGG16 질병 클래스
 */
export const VGG_CLASSES: ModelClassInfo[] = [
  { 
    id: 1, 
    name: '바이러스성출혈성패혈증', 
    nameEn: 'Viral Hemorrhagic Septicemia',
    description: '바이러스성 출혈성 패혈증',
    color: '#F44336',
    icon: 'drop'
  },
  { 
    id: 2, 
    name: '림포시스티스병', 
    nameEn: 'Lymphocystis Disease',
    description: '림포시스티스 바이러스 감염',
    color: '#FF9800',
    icon: 'virus'
  },
  { 
    id: 6, 
    name: '여윔병', 
    nameEn: 'Streptococcosis',
    description: '연쇄상구균 감염',
    color: '#4CAF50',
    icon: 'bacteria'
  },
  { 
    id: 8, 
    name: '스쿠티카병', 
    nameEn: 'Scuticociliatosis',
    description: '스쿠티카 섬모충 감염',
    color: '#9C27B0',
    icon: 'bug'
  },
  { 
    id: 11, 
    name: '연쇄구균증', 
    nameEn: 'Streptococcosis',
    description: '연쇄구균 감염',
    color: '#00BCD4',
    icon: 'bacteria'
  },
  { 
    id: 13, 
    name: '비브리오병', 
    nameEn: 'Vibriosis',
    description: '비브리오 균 감염',
    color: '#E91E63',
    icon: 'bacteria'
  },
  { 
    id: 19, 
    name: '에드워드병', 
    nameEn: 'Edwardsiellosis',
    description: '에드워드세균 감염',
    color: '#795548',
    icon: 'bacteria'
  },
];

/**
 * 기본 모델 설정
 */
export const DEFAULT_MODEL_CONFIG: ModelConfig = {
  yoloClasses: YOLO_CLASSES,
  vggClasses: VGG_CLASSES,
  inputTensorSize: [1, 3, 640, 640],
  confidenceThreshold: 0.5,
  iouThreshold: 0.45,
};

/**
 * 모델 클래스 유틸리티
 */
export class ModelUtils {
  /**
   * YOLO 클래스 이름 조회
   */
  static getYoloClassName(id: number, useEnglish = false): string {
    const cls = YOLO_CLASSES.find(c => c.id === id);
    return useEnglish ? cls?.nameEn || 'Unknown' : cls?.name || 'Unknown';
  }

  /**
   * VGG 클래스 이름 조회
   */
  static getVggClassName(id: number, useEnglish = false): string {
    const cls = VGG_CLASSES.find(c => c.id === id);
    return useEnglish ? cls?.nameEn || 'Unknown' : cls?.name || 'Unknown';
  }

  /**
   * VGG 클래스 정보 조회
   */
  static getVggClassInfo(id: number): ModelClassInfo | undefined {
    return VGG_CLASSES.find(c => c.id === id);
  }

  /**
   * 클래스 색상 조회
   */
  static getClassColor(id: number, isVgg = false): string {
    const classes = isVgg ? VGG_CLASSES : YOLO_CLASSES;
    return classes.find(c => c.id === id)?.color || '#999999';
  }

  /**
   * 클래스 아이콘 조회
   */
  static getClassIcon(id: number, isVgg = false): string {
    const classes = isVgg ? VGG_CLASSES : YOLO_CLASSES;
    return classes.find(c => c.id === id)?.icon || 'help-circle';
  }

  /**
   * 새로운 질병 클래스 추가
   */
  static addVggClass(classInfo: ModelClassInfo) {
    const existingIndex = VGG_CLASSES.findIndex(c => c.id === classInfo.id);
    if (existingIndex >= 0) {
      VGG_CLASSES[existingIndex] = classInfo;
    } else {
      VGG_CLASSES.push(classInfo);
    }
  }

  /**
   * 새로운 증상 클래스 추가
   */
  static addYoloClass(classInfo: ModelClassInfo) {
    const existingIndex = YOLO_CLASSES.findIndex(c => c.id === classInfo.id);
    if (existingIndex >= 0) {
      YOLO_CLASSES[existingIndex] = classInfo;
    } else {
      YOLO_CLASSES.push(classInfo);
    }
  }

  /**
   * 모든 VGG 클래스 ID 조회
   */
  static getVggClassIds(): number[] {
    return VGG_CLASSES.map(c => c.id);
  }

  /**
   * 클래스 매핑 정보 생성
   */
  static getClassMapping(isVgg = false): Record<number, string> {
    const classes = isVgg ? VGG_CLASSES : YOLO_CLASSES;
    return classes.reduce((acc, cls) => {
      acc[cls.id] = cls.name;
      return acc;
    }, {} as Record<number, string>);
  }
}

export default DEFAULT_MODEL_CONFIG;

File: fdm-tflite-detector/app/src/main/java/kr/re/etri/fdm/smartlens/config/ModelConfig.kt
package kr.re.etri.fdm.smartlens.config

/**
 * 모델 클래스 정보 데이터 모델
 */
data class ModelClassInfo(
    val id: Int,
    val name: String,
    val nameEn: String,
    val description: String? = null,
    val color: String,
    val icon: String
)

/**
 * 모델 설정 관리 클래스
 */
object ModelConfig {
    
    /**
     * YOLOv8 증상 클래스
     */
    val YOLO_CLASSES = listOf(
        ModelClassInfo(0, "Bleeding", "Bleeding", null, "#F44336", "drop"),
        ModelClassInfo(1, "Corrosion", "Corrosion", null, "#FF9800", "alert-circle"),
        ModelClassInfo(2, "Tumor", "Tumor", null, "#9C27B0", "circle-outline"),
        ModelClassInfo(3, "Ulcer", "Ulcer", null, "#E91E63", "alert-circle"),
        ModelClassInfo(4, "EyesSymptom", "Eye Symptom", null, "#2196F3", "eye")
    )
    
    /**
     * VGG16 질병 클래스
     */
    val VGG_CLASSES = listOf(
        ModelClassInfo(
            1, 
            "바이러스성출혈성패혈증", 
            "Viral Hemorrhagic Septicemia",
            "바이러스성 출혈성 패혈증",
            "#F44336",
            "drop"
        ),
        ModelClassInfo(
            2, 
            "림포시스티스병", 
            "Lymphocystis Disease",
            "림포시스티스 바이러스 감염",
            "#FF9800",
            "virus"
        ),
        ModelClassInfo(
            6, 
            "여윔병", 
            "Streptococcosis",
            "연쇄상구균 감염",
            "#4CAF50",
            "bacteria"
        ),
        ModelClassInfo(
            8, 
            "스쿠티카병", 
            "Scuticociliatosis",
            "스쿠티카 섬모충 감염",
            "#9C27B0",
            "bug"
        ),
        ModelClassInfo(
            11, 
            "연쇄구균증", 
            "Streptococcosis",
            "연쇄구균 감염",
            "#00BCD4",
            "bacteria"
        ),
        ModelClassInfo(
            13, 
            "비브리오병", 
            "Vibriosis",
            "비브리오 균 감염",
            "#E91E63",
            "bacteria"
        ),
        ModelClassInfo(
            19, 
            "에드워드병", 
            "Edwardsiellosis",
            "에드워드세균 감염",
            "#795548",
            "bacteria"
        )
    )
    
    // 모델 입력 텐서 크기
    const val INPUT_TENSOR_WIDTH = 640
    const val INPUT_TENSOR_HEIGHT = 640
    const val INPUT_TENSOR_CHANNELS = 3
    
    // VGG16 입력 크기
    const val VGG_INPUT_SIZE = 112
    
    // 신뢰도 임계값
    const val CONFIDENCE_THRESHOLD = 0.5f
    const val CONFIDENCE_THRESHOLD_LOW = 0.3f
    const val CONFIDENCE_THRESHOLD_HIGH = 0.8f
    
    // NMS IoU 임계값
    const val IOU_THRESHOLD = 0.45f
    
    /**
     * YOLO 클래스 이름 조회
     */
    fun getYoloClassName(id: Int, useEnglish: Boolean = false): String {
        return YOLO_CLASSES.find { it.id == id }?.let {
            if (useEnglish) it.nameEn else it.name
        } ?: "Unknown"
    }
    
    /**
     * VGG 클래스 이름 조회
     */
    fun getVggClassName(id: Int, useEnglish: Boolean = false): String {
        return VGG_CLASSES.find { it.id == id }?.let {
            if (useEnglish) it.nameEn else it.name
        } ?: "Unknown"
    }
    
    /**
     * VGG 클래스 정보 조회
     */
    fun getVggClassInfo(id: Int): ModelClassInfo? {
        return VGG_CLASSES.find { it.id == id }
    }
    
    /**
     * YOLO 클래스 정보 조회
     */
    fun getYoloClassInfo(id: Int): ModelClassInfo? {
        return YOLO_CLASSES.find { it.id == id }
    }
    
    /**
     * 클래스 색상 조회
     */
    fun getClassColor(id: Int, isVgg: Boolean = false): String {
        return (if (isVgg) VGG_CLASSES else YOLO_CLASSES)
            .find { it.id == id }?.color ?: "#999999"
    }
    
    /**
     * 클래스 아이콘 조회
     */
    fun getClassIcon(id: Int, isVgg: Boolean = false): String {
        return (if (isVgg) VGG_CLASSES else YOLO_CLASSES)
            .find { it.id == id }?.icon ?: "help-circle"
    }
    
    /**
     * 모든 VGG 클래스 ID 조회
     */
    fun getVggClassIds(): List<Int> {
        return VGG_CLASSES.map { it.id }
    }
    
    /**
     * 모든 YOLO 클래스 ID 조회
     */
    fun getYoloClassIds(): List<Int> {
        return YOLO_CLASSES.map { it.id }
    }
    
    /**
     * 질병 클래스 설명 조회
     */
    fun getDiseaseDescription(id: Int): String? {
        return VGG_CLASSES.find { it.id == id }?.description
    }
    
    /**
     * 새로운 질병 클래스 추가 (런타임)
     */
    fun addVggClass(classInfo: ModelClassInfo) {
        // 주의: 이 변경사항은 앱 재시작 시 초기화됨
        // 영구 저장을 위해서는 별도의 저장 메커니즘 필요
        val existingIndex = VGG_CLASSES.indexOfFirst { it.id == classInfo.id }
        if (existingIndex >= 0) {
            // 업데이트
            // 실제로는 불변 리스트이므로 새로운 리스트 생성 필요
        }
        // 새로운 클래스 추가는 ModelConfig 객체 재구성 필요
    }
}
